boards: renesas: Add support for EK-RA8D1, MCK_RA8T1

duynguyenxa opened the pull request:
This PR add support for EK-RA8D1 and MCK-RA8T1 based on the EK-RA8M1 development
Note: This should be merge after the https://github.com/zephyrproject-rtos/zephyr/pull/76312

duynguyenxa committed:
soc: renesas: Add support for RA8D1 SoC

Initial commit to suppor RA8D1 SoC
This is deveop base on RA8M1 so it will have similar stucture and
feature

Signed-off-by: Duy Phuong Hoang. Nguyen <duy.nguyen.xa@renesas.com>

duynguyenxa committed:
boards: renesas: Support Renesas EK-RA8D1 board

Initial support for EK-RA8D1 board this commit
only support basic GPIO and Serial driver for
RA8D1 board

Signed-off-by: Duy Phuong Hoang. Nguyen <duy.nguyen.xa@renesas.com>

duynguyenxa committed:
soc: renesas: add support for RA8T1 SoC

Initial commit to support RA8T1 SoC

Signed-off-by: The Nguyen <the.nguyen.yf@renesas.com>
Signed-off-by: Duy Phuong Hoang. Nguyen <duy.nguyen.xa@renesas.com>

duynguyenxa committed:
boards: renesas: add support for mck_ra8t1 board

Initial commit for MCK-RA8T1 board

Signed-off-by: The Nguyen <the.nguyen.yf@renesas.com>
Signed-off-by: Duy Phuong Hoang. Nguyen <duy.nguyen.xa@renesas.com>

soburi commented:
Please mark it as `draft` by the github feature.
![スクリーンショット 2024-07-02 061947](https://github.com/zephyrproject-rtos/zephyr/assets/458281/0daf8ef1-e9a3-480e-b71b-451ed92e0fa2)

soburi commented:
![スクリーンショット 2024-07-02 061947](https://github.com/zephyrproject-rtos/zephyr/assets/458281/6ef37dc6-fa9f-47c8-8c21-0dda4f0a3d41)
Please mark it as `draft` by the github feature.

soburi commented:
![スクリーンショット 2024-07-02 062558](https://github.com/zephyrproject-rtos/zephyr/assets/458281/dc06d160-7222-4010-8832-dea2bf6cc910)
Please mark it as `draft` by the github feature.

soburi commented:
And I think this PR is too huge.
Please split into each board addition and drivers addition.
It is better to add drivers first with it into RA8M1.

KhiemNguyenT commented:
> And I think this PR is too huge. Please split into each board addition and drivers addition. It is better to add drivers first with it into RA8M1.

@soburi this suggestion will be considered in next rebase. Perhaps, you will give us some comments on these patches, please.

KhiemNguyenT commented:
I have checked the patches. It's aligned with previous PR. Hope to receive some comments from others.

ydamigos commented on drivers/spi/spi_b_renesas_ra8.c:
```diff
@@ -0,0 +1,711 @@
+	return 0;
+}
+
+static const struct spi_driver_api ra_spi_driver_api = {.transceive = ra_spi_b_transceive,
+#ifdef CONFIG_SPI_ASYNC
+							.transceive_async =
+								ra_spi_b_transceive_async,
+#endif /* CONFIG_SPI_ASYNC */
+							.release = ra_spi_b_release};
```
```suggestion
static const struct spi_driver_api ra_spi_driver_api = {
	.transceive = ra_spi_b_transceive,
#ifdef CONFIG_SPI_ASYNC
	.transceive_async = ra_spi_b_transceive_async,
#endif /* CONFIG_SPI_ASYNC */
	.release = ra_spi_b_release
};
```

ydamigos commented on drivers/spi/spi_b_renesas_ra8.c:
```diff
@@ -0,0 +1,711 @@
+#if CONFIG_SPI_B_INTERRUPT
+	uint32_t data_len;
+#endif
+#if defined(CONFIG_SPI_B_RA_DTC)
```
Is it possible for someone to use DTC only on RX or TX?
If yes, we should use tx-dtc and rx-dtc DT properties.
If no, should we consider use only one property for DTC in DT (e.g. use_dtc)?

duynguyenxa replied:
@ydamigos , User are able to to use only tx dtc or rx dtc.
`#ifndef CONFIG_SPI_B_RA_DTC
#define RA_SPI_B_DTC_STRUCT_INIT(index)
#define RA_SPI_B_DTC_INIT(index)
#else
#define RA_SPI_B_DTC_INIT(index)                                                                   \
	do {                                                                                       \
		if (DT_INST_PROP_OR(index, rx_dtc, false)) {                                       \
			ra_spi_data_##index.fsp_config.p_transfer_rx =                             \
				&ra_spi_data_##index.rx_transfer;                                  \
		}                                                                                  \
		if (DT_INST_PROP_OR(index, tx_dtc, false)) {                                       \
			ra_spi_data_##index.fsp_config.p_transfer_tx =                             \
				&ra_spi_data_##index.tx_transfer;                                  \
		}                                                                                  \
	} while (0)`

soburi commented on tests/drivers/i2c/i2c_api/src/test_i2c.c:
```diff
@@ -28,6 +28,16 @@
 
 uint32_t i2c_cfg = I2C_SPEED_SET(I2C_SPEED_STANDARD) | I2C_MODE_CONTROLLER;
 
+#define HMC_ADDR (0x1E)
```
Please define in Kconfig and use `extra_configs` in testcase.yaml.

duynguyenxa replied:
@soburi , It's updated please have a look

soburi commented on soc/renesas/ra/ra8m1/soc.c:
```diff
@@ -17,10 +17,14 @@
 
 #include <bsp_api.h>
 
+#define CCR_CACHE_ENABLE (0x000E0201)
```
Can this value be expressed in the cmsis definition?

duynguyenxa replied:
@soburi , I changed it into
`/* Enable Instrcution cache, Branch prediction, Loop and branch info cache */
#define CCR_CACHE_ENABLE (SCB_CCR_IC_Msk | SCB_CCR_BP_Msk | SCB_CCR_LOB_Msk)`

soburi commented on drivers/clock_control/clock_control_renesas_ra_cgc.c:
```diff
@@ -0,0 +1,90 @@
+
+static int clock_control_renesas_ra_on(const struct device *dev, clock_control_subsys_t sys)
+{
+	struct clock_control_ra_subsys_cfg *subsys_clk = (struct clock_control_ra_subsys_cfg *)sys;
```
Please do not cast `clock_control_subsys`_t to different types ​​for on, off, and get_rate. For places where the same type is used as an argument in the clock_control API, please use a consistent type according to the definition of the original API.

duynguyenxa replied:
@soburi , I refer to some other clock control driver, they also do casting for subsys clock.

e.g. https://github.com/zephyrproject-rtos/zephyr/blob/fd0e46506fabd4c0c91ecb336a39cb6a6aa2b270/drivers/clock_control/clock_stm32_ll_u5.c#L151C9-L151C22

Please correct me if I'm wrong, as my understanding in the description in clock_control.h, the clock_control_subsys_t is designed to be casted to the data structure of sub-system driver instance being used.

/**
 * clock_control_subsys_t is a type to identify a clock controller sub-system.
 * Such data pointed is opaque and relevant only to the clock controller
 * driver instance being used.
 */
typedef void *clock_control_subsys_t;

soburi replied:
Sorry, I checked something incorrectly, so I'll comment again.

Here, casting itself is not a problem.
The type is defined as void*, so it can be any type.

By the way, how is the clock_control_subsys_t argument processed in get_rate?

Here, the function is expected to be called with the same clock_control_subsys_t value as in the on/off case, so the corresponding clock frequency is expected to be obtained.

In other words, if the same value as a clock_control_subsys_t that allows on and off to operate normally is passed to get_rate, that frequency is expected to be obtained for this. We think it should be possible to get the pclk[abc...] value corresponding to each peripheral, not the system clock.

duynguyenxa replied:
@soburi -san, this implement treat each pclk as a clock device, hence it will get the rate of pclk[abc...] by the device config. Similar as the implementation of fixed clock driver in Zephyr

soburi replied:
In Zephyr, it is expected to call the `get_rate` of clock_controller with the value of `clocks` for each node set in the device tree to obtain the clock frequency required by that peripheral. The API of clock_controller is designed with this usage in mind.

In the case of Renesas RA, the FSP mechanism can be used to obtain the clock information required by the driver without using this method,  so this is not a problem at a glance, but this is the FSP way and not Zephyr's. Since clocks affect the entire system, they should be designed appropriately according to Zephyr's custom; the get_rate of clock_controller should be implemented appropriately.

For example, the SCI uses pclka in RA4M1 and the pclkb in RA2A1.
I hope it can switch between these in the device tree description and obtain the frequency of each clock through get_rate.

In practice, I think adding one cell to specify the type of clock is enough.

duynguyenxa commented on drivers/spi/Kconfig.renesas_ra8:
```diff
@@ -0,0 +1,33 @@
+
+config SPI_B_RA_DTC
+	bool "RA MCU SPI DTC Support"
+	select USE_RA_FSP_DTC
```
@ydamigos, As for now it's seemed there is no good way to get the property value of rx-dtc and tx-dtc without the node name so I will set this enable as default.

duynguyenxa committed:
manifest: Update commit ID to support new renesas hal

This update is to support new version of renesas hal

Signed-off-by: Duy Nguyen <duy.nguyen.xa@renesas.com>

duynguyenxa commented:
@soburi ,  As you requested earlier, we separate this PR into smaller chunk of driver support for RA8 series, this PR is now only to add RA8D1 and RA8T1 support based on the existed RA8M1.

duynguyenxa committed:
tests: Skip CPP98 for Renesas device using FSP hal

As Renesas Hal is not compatitble with C++98 we use
filter CONFIG_HAS_RENESAS_RA_FSP to skip the test

Signed-off-by: Duy Nguyen <duy.nguyen.xa@renesas.com>

ydamigos commented on boards/renesas/ek_ra8d1/ek_ra8d1.yaml:
```diff
@@ -0,0 +1,11 @@
+ram: 1024
+flash: 2048
+toolchain:
+  - gnuarmemb
```
we could also add zephyr here, couldn't we?

duynguyenxa replied:
@ydamigos , I added it

soburi reviewed (dismissed):
LGTM

ydamigos commented on west.yml:
```diff
@@ -214,7 +214,7 @@ manifest:
     - name: hal_renesas
       path: modules/hal/renesas
-      revision: af77d7cdfeeff290593e7e99f54f0c1e2a3f91e6
+      revision: pull/24/head
```
PR was merged in hal_renesas. Please use latest hash.

ydamigos reviewed (changes requested):
west.yml needs to be updated.

duynguyenxa committed:
driver: serial: Convert Kconfig of renesas RA8 to rsource

Change Kconfig of renesas RA8 to rsource

Signed-off-by: Duy Nguyen <duy.nguyen.xa@renesas.com>
